provider/kubernetes/kubernetes.go

```diff
@@ -159,10 +140,66 @@ func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *s
 			log.Errorf("Cannot connect to Provider: %s", err)
 		}
 	})
-
 	return nil
 }
 
+func (p *Provider) watch(k8sClient Client, configurationChan chan<- types.ConfigMessage) error {
+	// This controller watches the events channel, and allows the client watches to be restarted if needed
+	stopWatchNS := make(chan struct{}, 1)
+	defer close(stopWatchNS)
+
+	namespaceChan, err := k8sClient.WatchNamespaces(p.Namespaces, stopWatchNS)
+	if err != nil {
+		log.Errorf("Error watching kubernetes namespace events: %s", err)
+		return err
+	}
+
+	for {
+		//	Start watching the namespaced content
+		err := p.watchNamespaceContent(k8sClient, configurationChan, namespaceChan)
+		if err != nil {
+			return fmt.Errorf("error watching namespaced content: %v", err)
+		}
+	}
+}
```

Comment:
I'm concerned about performance/scalability here: AFAIU, a change in one of the watched namespaces (addition, removal, etc.) causes all informers for all watched namespaces to be recreated. We have to wait for each informer to be synched (i.e., populated), and we do so sequentially, which means that no Ingresses can be processed during this period.

The delays incurred could become unacceptable for installations with a larger number of namespaces, or those with a moderate number of namespaces that contain a large number of objects. Aggravatingly, the resync period would cause a reset to happen even when no watched namespaces change. (That issue could be fixed rather easily though by disabling the resync period just for the namespace watcher.)

Given my understanding is correct, I believe that in order for the approach to become bearable, we would have to change the implementation such that we identify delta changes happening on the namespaces watch (i.e., distinguish between namespaces getting added, removed, or modified) and respond to those only. That is, we start watching new namespaces and remove watches on disappearing namespaces, all while keeping unchanged namespaces untouched. But even then, we might still have to run these optimized changes concurrently so that Ingress processing for untouched namespaces does not slow down.

I wonder if the added complexity we'd have to buy here still justifies the feature value. Before considering that path I'd like to hear what you have to say first -- my hope is that I get things wrong and everything is actually much better than I see it right now. :-)

Reply:
I think that you are correct about performance and scalability. The delay could potentially scale out of the trivial range, but in environments where scalability/performance is an issue, the dynamic namespace addition may not be a feasible solution, and they may need to stick to the static selection.

The namespace watcher is on a completely separate channel, and should not trigger on anything other than a namespace event.

You are correct about having to implement a delta system for watches for large numbers of objects, but of course, due to synchronicity issues and caching, there is a possibility that the namespace may not meet the requirements before it is removed, and may cause issues, even with the statically defined namespace filter that we previously used.

The point of this PR was to change as little of the operational code as possible (merely change the flow to extend functionality), and then potentially optimize in the future if performance issues arise (as described above, there should be no performance change for statically defined namespaces). This way, the feature could be released in 1.7, and then improved in future patch releases, as needed, without holding up the feature for smaller environments where performance may not be an issue.

Reply:
>This way, the feature could be released in 1.7, and then improved in future patch releases, as needed, without holding up the feature for smaller environments where performance may not be an issue.

I'm not too sure about this. These days, Traefik seems to be used in many environments where scalability is needed right from the start, with thousands of Ingresses involved sometimes. The current Ingress processing implementation is very fast and responds to updates rapidly, which I think users appreciate and possibly even expect at this point. My feeling is that if we don't build in scalability right from the start for new features, we risk losing more users than we can win by a functional but under-performing functionality.

I think it'd help a lot if we knew the approximate point at which the implementation would start degrading measurably. I'd be much more relaxed if we knew that namespace changes for 500 Ingresses can be handled just fine, but not so much if it were 50. Maybe we can try to quantify the impact?

>The namespace watcher is on a completely separate channel, and should not trigger on anything other than a namespace event.

Except for the resync period which causes "fake" events to be emitted periodically. At the current state, we [reuse the existing resync period](https://github.com/dtomcej/traefik/blob/837f2548c07ca0dab3325127bafbaadc23818e81/provider/kubernetes/client.go#L122) so informer factories are recreated at least every 10 minutes. This could be a lightweight operation or induce seconds-scale latencies, depending on the number of namespaces and objects.